Cleanup package builds and config

favna opened the pull request:
- Upgrade all outdated dependencies.
  - Tip: Use something like depfu.com to automate this in the future
- Reconfigure jest so the configuration doesn't muddy up the package.json
- Add jest-runner-tsc to the main `jest` stack... it would be quite weird to not test your own code with your lib when writing a test runner.
- Restructure the files and folders a bit to be more concise

favna committed:
Cleanup package builds and config

- Upgrade all outdated dependencies. Tip: Use something like depfu.com to automate this in the future
- Reconfigure jest so the configuration doesn't muddy up the package.json
- Add jest-runner-tsc to the main `jest` stack... it would be quite weird to not test your own code with your lib when writing a test runner.
- Restructure the files and folders a bit to be more concise

Signed-off-by: Jeroen Claassens <jeroen.claassens@live.nl>

favna commented:
Note that this will also massively decrease the files published to NPM as it should be. This is the new dry-run:

```
$ npm publish --dry-run
npm notice
npm notice package: jest-runner-tsc@0.0.1-development
npm notice === Tarball Contents ===
npm notice 1.1kB package.json
npm notice 1.1kB LICENSE
npm notice 1.0kB README.md
npm notice 157B  dist/index.js
npm notice 4.1kB dist/runTsc.js
npm notice === Tarball Details ===
npm notice name:          jest-runner-tsc
npm notice version:       0.0.1-development
npm notice package size:  3.1 kB
npm notice unpacked size: 7.5 kB
npm notice shasum:        9dd6a94e4863c12b4cd998b82d22547e8c9c9740
npm notice integrity:     sha512-Q7ZLwAAgLo358[...]ciS4AlYXtVSKQ==
npm notice total files:   5
npm notice
+ jest-runner-tsc@0.0.1-development
```

While this is the current publish:

![image](https://user-images.githubusercontent.com/4019718/51941365-b2401200-2414-11e9-92a3-7632c829d839.png)

favna committed:
Fix travis

Apparently the upgraded devDependencies don't run on NodeJS 6 or 9 (sane
has a very weird engine specification) but considering that NodeJS 10 is
LTS, 11 is current and 8 is still supported just doing it this way
should be fine

Signed-off-by: Jeroen Claassens <jeroen.claassens@live.nl>

Favna commented:
WIP: Do not merge

Apparently semantic-release has completely changed since the version you were used so this needs fixing

favna committed:
Fix semantic-release

azz commented on test/fixture/bad.ts:
```diff
@@ -1,3 +0,0 @@
-function bad(): string {
```
Why delete these fixtures?

favna replied:
Because there is no way to tell this lib something should fail by design and this is, by design, bad TypeScript code. Including this causes an error in the tests:

```sh
➜  jest-runner-tsc git:(master) ✗ yarn test
yarn run v1.13.0
$ npm run-script build && jest

> jest-runner-tsc@0.0.0-development build /Users/jeroenc/dev/jest-runner-tsc
> babel src -d dist

Successfully compiled 2 files with Babel.
 PASS   lint  src/index.js
 PASS   lint  ./jest.config.js
 PASS   lint  test/index.test.js
 PASS   lint  src/runTsc.js
 FAIL   tsc  test/lib.test.ts
Type '42' is not assignable to type 'string'.
   7 | describe('compiling incorrect TypeScript should fail', () => {
   8 |   const libTestFail = (): string => {
>  9 |     return 42;
     |     ^^^^^^^^^^
  10 |   }
  11 | });
 PASS   test  test/index.test.js

Test Suites: 1 failed, 5 passed, 6 total
Tests:       1 failed, 5 passed, NaN total
Snapshots:   0 total
Time:        6.436s
Ran all test suites in 3 projects.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

This is after modifying `lib.test.ts` to:

```ts
describe('compiling correct TypeScript should work', () => {
  const libTest = (): string => {
    return 'apples';
  };
});

describe('compiling incorrect TypeScript should fail', () => {
  const libTestFail = (): string => {
    return 42;
  }
});
```

So unless the lib gets some feature where we can tell Jest that a test should fail it is not possible to test the lib with its own runner while also having a test that should intentionally fail.

This would also happen before when you'd run `jest -c jest.tsc.config.js`, however now that I have made it so all runners are always ran it happens always.

Note that this therefore assumes that the lib itself also runs over its own test code which I personally thing it really should do. Just look at projects like TypeScript that's written entirely *in* TypeScript or Webpack which packs itself with webpack.

azz replied:
Can we simply add them to ignore paths in jest config?

favna replied:
Sure we can but then that would be the same end goal as excluding entirely. I mean ignored tests are basically useless. Sure it might be a nice example for other people, but I'm quite sure that anyone starting TypeScript wouldn't make such a mistake let alone need an example in this lib to have them understand it's wrong 

(Considering both running tsc or ts-node will crash on this type of bad code)

azz replied:
This test was to check that this test runner actually failed when tsc failed. Without it we have no guarantee it's actually doing its job.

azz commented on package.json:
```diff
@@ -4,54 +4,37 @@
+    "jest": "^24.0.0",
+    "jest-runner-eslint": "^0.7.1",
+    "prettier": "^1.16.1",
+    "semantic-release": "^15.13.3",
```
Sure this change doesn't require reconfiguring `.travis.yml`? I've had that problem in the past.

favna replied:
Are you talking about the upgrade to semantic-release?

Honestly I... wouldn't know. I never used it any of my own packages. I could revert it if you mean that one.

azz replied:
Yeah might be best to leave `semantic-release` at v8 for now.

azz commented on .npmignore:
```diff
@@ -0,0 +1,14 @@
+.git/
```
I don't like using `.npmignore` files. I'd much prefer using `files` property in `package.json`  to be more explicit about what is included in the package. 

https://medium.com/@jdxcode/for-the-love-of-god-dont-use-npmignore-f93c08909d8d

favna replied:
I'll make this change when I'm home

azz commented on .gitignore:
```diff
@@ -1,3 +1,3 @@
 node_modules
-*.log
 dist
+.idea/
```
You should use something like [this](https://help.github.com/articles/ignoring-files/#create-a-global-gitignore) instead. I keep `*.log` in all my `.gitignore`s to catch `yarn-error.log` etc.

favna replied:
.idea folder is for the IDE IntelliJ IDEA much like how VSCode created a .vscode folder. There are plenty of projects that I personally work on (heck the company I work for in fact) where the .idea folder does get comitted, it is therefore unrealistic to globally ignore it. 

I could remove it yes but that does end up meaning that if other people fork your project they are constantly going to have to either pass that folder when `git add`ing or delete it entirely before said process.

azz replied:
Ok no worries :+1:

azz commented on .babelrc:
```diff
@@ -0,0 +1,5 @@
+{
```
Why revert back to `.babelrc`?

favna replied:
3 Reasons

1. It is the [standard for babel when not wanting to compile node_modules (which we don't)](https://babeljs.io/docs/en/configuration)
2. Support for plugins / build in features that IDE's such as VSCode and IntelliJ IDEA have to recognise it as a babel configuration and provide proper IntelliSense
3. Support for plugins such as Material Icons in common IDE's to recognise it as a Babel file. Visual only but nice regardless.

azz replied:
Ok sounds good.

azz reviewed (dismissed):
Thanks for updating things! I left a few inline comments.

favna commented:
@azz Thanks for the feedback! I got back to you on all remarks you made. Lets figure this out together.

Commit:
Restore failing test but ignore in main run

Commit:
Downgrade semantic-release

Commit:
Use files instead of npmignore

favna commented:
@azz I've made the requested changes. If you could review and dry run semantic-release that would be great.

azz commented on package.json:
```diff
@@ -4,57 +4,41 @@
   "scripts": {
     "build": "babel src -d dist",
-    "test": "jest",
+    "test": "npm run-script build && jest",
```
`yarn build` also works.

Commit:
More small package tweaks

- Run build script eith yarn in test - since this project relies on yarn
anyway
- version spec create-jest-runner. They haven't updated their changelog
yet so there is no telling if 0.5.0 would be breaking

Signed-off-by: Jeroen Claassens <support@fanva.xyz>

azz commented:
Thanks. I was just reviewing the state of master (been a while since I looked at this package) and realised that the "bad" fixtures were _never_ automatically verified. I must have been doing it manually... So your changes are perfectly fine.

azz commented on package.json:
```diff
@@ -4,57 +4,41 @@
   "scripts": {
     "build": "babel src -d dist",
-    "test": "jest",
+    "test": "yarns build && jest",
```
typo: `yarn`

favna replied:
Woops! Good thing we have TravisCI to catch these things too.

Commit:
Fix Typo

azz reviewed (approved):
LGTM. Sorry for being slow on the reviews!

azz commented:
Automatically published as `1.4.0`, feel free to try it out to sanity-check.

favna commented:
No prob! A final remark about that last commit where I version specced create-jest-runner, I [created an issue](https://github.com/jest-community/create-jest-runner/issues/15) asking about the 0.5.0 changelog, though as I say there too I was also just very fast to notice it since release coz I just so happened to run `yarn outdated`. If nothing major changed I suppose that package can get updated too. (From `~0.4.1` to `^0.5.0`)


(btw you're still way faster than some other repos I contributed to lmao)
